Switch MapBlock compression to zstd

lhofhansl opened the pull request:
This add zstd to Minetest.

The code backward, but not forward compatible:
- blocks are compressed with zstd in the wire if the client understands, otherwise zlib is used as before
- all new blocks written to the DB are compressed with zstd
- old blocks are still decompressed with zlib

@sfan5's brotli and @est31's earlier zstd branched helped a lot - although I'm using different zstd APIs.

Questions:
- What about the compression levels? Right now these are zlib specific. For now I simply mapped them to zstd levels 1-10, which really should be good enough. Alternatively I could remove the compression level config, since zstd's default is pretty good.
- Is forward compatibility needed? I.e. should we (somehow) allow old server to still read old maps, even if a newer server has written to them?
- What should the default compression level be? Right now I set it to 0 (which uses the default, which currently is 3). That seems pretty good. Better compression than zlib, and faster too.

**Some stats: (scene with viewing_range of 600, 3700 blocks visible)**
**Zlib-6 (default)**, 54MB DB, 90s to generate (dominated by compression time on save)
**Zlib-3**, 69MB DB, 70s to generate

**Zstd-15**, 50MB, 108s to load
**Zstd-10**, 50MB, 70s to load
**Zstd-6**, 54MB, 67s to load
**Zstd-3 (default)**, 56MB DB, 60s to load
**Zstd-1**, 55MB DB, 60s to load
**Zstd- -1**, 62MB, 60s to load
**Zstd- -10**, 72MB, 60s to load

This shows - as expected - that we get more compression per time from zstd. It also shows that for the tiny things like blocks higher zstd level do not buy much. Perhaps we should default to 1. I'm concluding that zstd below level 3 is no longer the bottelneck. (should confirm with a run with no compression)
For zlib levels other than 3 or 6 do not make much sense, while for zstd level 1 seems to be optimal.

See @ShadowNinja's comment below for more ways to optimize this.

I sporadically watched the map save times. Grew up to 5s for zlib-6, stayed below 2s for zlib-3.
Zstd-10 stayed below 2s, Zstd-1 below 1s.
Similarly Zstd-15 stayed below 5s. Not much improvement there for levels -1 or -10.

~~(I did not test zstd compression levels larger than 10, or negative values - which are faster but with less compression)~~

lhofhansl committed:
Add zstd support.

lhofhansl commented on src/serialization.cpp:
```diff
@@ -197,11 +198,98 @@ void decompressZlib(std::istream &is, std::ostream &os, size_t limit)
+	if(version >= 29)
+	{
+		// map the zlib levels [-1,9]
+		// the range 1-10 is good enough, 0 is the default (currently 3)
```
You may find this questionable.
Perhaps, since compress can now call multiple schemes, we ought to come with a general compression scale, perhaps -1 to 100, and then map that whatever range the specific algorithms support.

Or remove the config option altogether and use the defaults....

lhofhansl replied:
Yet another option is to always use -1 for zlib, and use the zstd spectrum for the config instead.
zstd has a wide range compressions from very fast to very slow, and perhaps we'll never need a new one again.

lhofhansl commented:
Oops. Removed some unrelated changes.

lhofhansl commented on src/unittest/test_compression.cpp:
```diff
@@ -174,6 +176,42 @@ void TestCompression::testZlibLargeData()
+	infostream << "Test: Testing zstd wrappers with a large amount "
+		"of pseudorandom data" << std::endl;
+
+	u32 size = 500000;
```
500.000 because zstd default buf size is 128k.

lhofhansl replied:
No longer needed now since I changed the bufsize to 64, but it's still a good test, and takes only a few ms anyway.

lhofhansl commented:
I can clearly use some help with the build issues. It builds fine on my Linux box.

lhofhansl commented:
Looks like `ZSTD_c_compressionLevel` is not defined in older version of zstd, neither ZSTD_e_end, which is need here. Hmmm.

lhofhansl commented:
OK. At least it now compiles with the legacy API used...
I have no clue why the unittest fails. It passes on my machine, but I have a newer version of libzstd.

lhofhansl commented:
Hah. Older versions of zstd require to initialize a dstream before each decompression. New version do not require that.

ShadowNinja commented:
I was actually looking into implementing this myself, but didn't finish it.  Some discussion on other optimizations to the MapBlock format: #3946

Something that might be worth looking into with zstd would be a custom dictionary -- it should improve compression on the small bits that we pass to the compressor.  We compress each part of the mapblock separately, so the inputs are generally 8KiB or less, which isn't much for a compressor to learn from without a precomputed dictionary.  It might also help to compress an entire mapblock instead of compressing nodes, meta, etc separately.  That might also improve performance since you only have to include the compressor startup time once (although I think zstd has some significant optimizations in startup time over zlib).

I wrote a C++ app that converted Minecraft worlds to a Minetest database, profiled it, and found that it spent most of its CPU time in zlib compressing Minetest blocks, so compression performance is definitely important to database performance.

lhofhansl commented:
Thanks @ShadowNinja. There's also #10671 (you've seen it). It turns out we write a lot of blocks to disk when the only thing changed is the timestamp.

I've improved startup somewhat by making the stream context thread local.

A pre-computed dictionary is tricky... The data can be only decompressed with that exact same dictionary (unless it is written with the data).

In addition to compression I find that ostringstream is unreasonably slow - especially construction, reducing the number of those will also help a lot.

lhofhansl commented:
I do not know how to fix the Android and Windows builds. It seems we're downloading pre-compiled artifacts, I do not have the necessary environments to build one for libzstd.

lhofhansl commented:
@ShadowNinja See second commit. For version >= 29 node data and metadata is now compressed together. Can't say I like the code - littered with if-statements, but it works.

ShadowNinja commented:
Yeah, I ended up using a `Buffer<u8>` to avoid the overhead of `std::ostringstream`.

I've pushed what I had completed here: https://github.com/ShadowNinja/minetest/commits/mapblock-v29

Only the serialization is completed, and there are a bunch of other changes on top of the compression, but it might be useful.

lhofhansl commented:
I like that better. (BTW, as I found out the hard way, you cannot be use the new compress2 zstd API if you want this work with zstd 1.3.3.)
Do you want to continue with your branch, or should I incorporate this in my branch. Fine either way, no ego from my side :)

ShadowNinja commented:
You should continue with your branch.  Mine is far from complete, and I'm not working on it right now.

lhofhansl commented:
Re: Returning a buffer from seralizeBulk, won't that call the copy constructor and make another copy of the buffer? (At least on some platforms since NRVO is not guaranteed.)

lhofhansl commented:
I removed the node data/metadata combined compression from this patch, since I did not like it.
Perhaps it's best to add zstd first, and then optimize other things around it.

lhofhansl commented:
@ShadowNinja new version, that compresses the entire block.

lhofhansl committed:
Compress entire mapblock

lhofhansl committed:
Fix Windows build.

lhofhansl commented:
The VS builds at least compile now. Still a linker problem.

lhofhansl commented:
The VCPKG builds now are fine. That's all I am going to be able to do, I'll need some help with the buildbot and Android - both of which load some pre-compiled binaries as far as I can see. @sfan5 ?

lhofhansl commented on src/serialization.cpp:
```diff
@@ -197,11 +198,118 @@ void decompressZlib(std::istream &is, std::ostream &os, size_t limit)
+
+	ZSTD_initCStream(stream.get(), level);
+
+	const size_t bufsize = 16384;
```
Earlier versions used `ZSTD_CStreamOutSize()`, but dynamic arrays are not C++ standard. I measure this performance with typically input, and found no measurable performance difference. I prefer this over repeated heap allocations.

lhofhansl commented:
OK. Docker work too. Now I need to figure out how to get a prebuild zstd image onto minetest.kitsunemimi.pw (and MinGW builds). And the same for Android.

lhofhansl commented:
I tested the following scenarios:
- server with zstd loading an old world -> works
- old server trying to load a new world with some blocks written in zstd -> fails, but is expected
- connecting an old client to a new server -> works
- connecting a new client to an old server -> works

lhofhansl commented:
In case it's not clear from above... I'm not planning any further work on this until there's a review (and someone points me to how I can get the pre-compiled libraries I need to Windows and Android).

sfan5 commented:
I plan to review this when time permits but I can answer those questions:
* For win32/win64 I will build a library and upload it to the place where the other libraries are
* For Android a build scripts needs to be added to https://github.com/minetest/minetest_android_deps, then someone needs to run that and push the result to https://github.com/minetest/minetest_android_deps_binaries

lhofhansl commented:
Thanks @sfan5 . I guess my above message missed a smiley somewhat. Upon re-reading it sounds a bit whiny.

ExeVirus commented:
Zstd is under BSD license, meaning we could static link it into minetest, meaning you don't need pre-compiled windows or android support. We could even do a .git submodule so we always use the latest release for builds. It's pure C, so I see no issues for any given platform.  Long story short: why are we not compiling this with minetest and keeping it a separate package?

sfan5 commented:
While that would be convenient we avoid vendoring other libraries too. It can be considered if zstd turns out to be too hard to install on platforms we support, though I doubt that is the case.

ExeVirus commented:
Okay, no worries. It was just a suggestion, in that case, the precompiled windows .dlls are here: https://github.com/facebook/zstd/releases/tag/v1.4.8

As for android, lemme look around, I'll update this comment with a link if I find it.

lhofhansl commented:
Thanks @ExeVirus . I was thinking about this too, but ultimately think we should avoid that as best as we can. We'd have to track updates ourselves, etc.
Not sure how much we can trust externally built libraries. The zstd ones are probably OK.

lhofhansl commented:
Time to pick this up again for 5.5?

ExeVirus commented:
Absolutely! What needs to be done? I can help a few hours a week till it's
merged. I know we need to do android deps, what else? Windows testing?

On Sun, Feb 28, 2021, 5:25 PM lhofhansl <notifications@github.com> wrote:

> Time to pick this up again for 5.5?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/minetest/minetest/pull/10788#issuecomment-787535481>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKT7N2XQLCFOUDPGFXTYRGDTBK7GBANCNFSM4VWQVMSA>
> .
>

lhofhansl commented:
Yeah. Need to build a Android and Windows zstd lib. I do not have environments (and time, frankly :( ) for either.

lhofhansl commented:
If we're still considering this, I'll rebase.

SmallJoker commented:
Is a hard dependency on zsed required or could this be modular?

sfan5 commented:
Here's the builds you were waiting for:
* `http://minetest.kitsunemimi.pw/zstd-1.4.9-win32.zip` (can be integrated with buildbot)
* `http://minetest.kitsunemimi.pw/zstd-1.4.9-win64.zip`
* https://github.com/minetest/minetest_android_deps_binaries/pull/5 (probably not testable without android build env)

I'll review the PR itself hopefully soon.
I also think that repeatable benchmarks would be useful here (doesn't need to be in final version). I'm thinking of command line flags that 1) load each map block and time decompression 2) recompress (load + write again) all mapblocks and time compression. You can find inspiration for this in the older issue.

lhofhansl commented:
Rebased. Let's wait for another test run.

lhofhansl committed:
Docker fix

lhofhansl commented:
@sfan5 You added benchmark for some PR before, right? Could you point me to that?

lhofhansl committed:
buildwin fixes

lhofhansl commented:
We still have to work out and agree upon the compression levels. Should we use the zstd levels (which have different meanings from the zlib levels)?

lhofhansl commented:
So far so good. Only Android is left now!

sfan5 commented:
> @sfan5 You added benchmark for some PR before, right? Could you point me to that?

https://github.com/sfan5/minetest/commit/062fca53365b6cf79097445c1476b87615f24977
https://github.com/sfan5/minetest/commit/b42a124c8416a32f71e0ba4e1f47044f1b9cb52b
https://github.com/sfan5/minetest/commit/8f9eeadc1cf9d820ee9162447dc609b389fa97bf

lhofhansl commented:
Thanks @sfan5

lhofhansl commented:
I'm removing the WIP label. I've been run my local server (just two people, though), with this for many, many weeks, without seeing an issues.

sfan5 commented on builtin/settingtypes.txt:
```diff
@@ -1045,11 +1045,10 @@ full_block_send_enable_min_time_from_building (Delay in sending blocks after bui
-#    ZLib compression level to use when sending mapblocks to the client.
-#    -1 - Zlib's default compression level
-#     0 - no compresson, fastest
+#    compression level to use when sending mapblocks to the client.
```
```suggestion
#    Compression level to use when sending mapblocks to the client.
```

sfan5 commented on builtin/settingtypes.txt:
```diff
@@ -1248,12 +1247,11 @@ max_objects_per_block (Maximum objects per block) int 64
-#    ZLib compression level to use when saving mapblocks to disk.
-#    -1 - Zlib's default compression level
-#     0 - no compresson, fastest
+#    compression level to use when sending mapblocks to the client.
```
```suggestion
#    Compression level to use when saving mapblocks to disk.
```

sfan5 commented on src/CMakeLists.txt:
```diff
@@ -271,9 +271,13 @@ if(WIN32)
 	find_library(ZLIB_LIBRARIES "zlib" DOC "Path to zlib library")
 
+	find_path(ZSTD_INCLUDE_DIR "zstd.h" DOC "Zstd include directory")
+	find_library(ZSTD_LIBRARY "zstd" DOC "Path to zstd library")
```
why does Windows do this differently and not use the CMake module? :thinking:
not important though, can be changed later (along with Zlib, there's no reason for it do this)

lhofhansl replied:
Yeah... I followed what zlib was doing.

sfan5 commented on src/map.h:
```diff
@@ -31,6 +31,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "voxel.h"
 #include "modifiedstate.h"
 #include "util/container.h"
+#include "util/pointer.h"
```
uh?

lhofhansl replied:
Might have been a left-over from some earlier changes.

sfan5 commented on src/mapblock.cpp:
```diff
@@ -494,15 +524,20 @@ void MapBlock::deSerialize(std::istream &is, u8 version, bool disk)
+		if (version >= 29) {
+			m_node_metadata.deSerialize(is, m_gamedef->idef());
+		} else {
+			// resuse iss
+			iss.str(std::string());
```
```suggestion
			// reuse iss
			iss.str("");
```

sfan5 commented on src/serialization.cpp:
```diff
@@ -22,6 +22,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "util/serialize.h"
 
 #include "zlib.h"
+#include "zstd.h"
```
```suggestion
#include <zlib.h>
#include <zstd.h>
```

sfan5 commented on src/serialization.cpp:
```diff
@@ -197,11 +198,118 @@ void decompressZlib(std::istream &is, std::ostream &os, size_t limit)
+		}
+	} while (ret != 0);
+
+	// Unget all the data that inflate didn't take
```
```suggestion
	// Unget all the data that ZSTD_decompressStream didn't take
```

sfan5 commented on src/serialization.cpp:
```diff
@@ -197,11 +198,118 @@ void decompressZlib(std::istream &is, std::ostream &os, size_t limit)
 {
+	if(version >= 29)
+	{
+		// map the zlib levels [-1,9]
+		// the range 1-10 is good enough, 0 is the default (currently 3)
```
was a bit confusing so I suggest this reword:
```suggestion
		// map the zlib levels [0,9] to [1,10]. -1 becomes 0 which indicates the default (currently 3) 
```

lhofhansl replied:
I wasn't sure whether to do it this way at all. Rather than mapping the zlib levels, perhaps we should allow the full range of zstd levels...?

sfan5 replied:
Yeah that'd be good. Allow the full range for zstd, cap the setting to 9 if zlib is used.

sfan5 commented on src/serialization.cpp:
```diff
@@ -240,8 +348,20 @@ void compress(const SharedBuffer<u8> &data, std::ostream &os, u8 version)
+void compress(const std::string &data, std::ostream &os, u8 version, int level)
+{
+	SharedBuffer<u8> databuf((u8*)data.c_str(), data.size());
+	compress(databuf, os, version, level);
```
It'd be better to move the implementation into here and make `compress(const SharedBuffer<u8>&, ...)` call this one because it'd avoid needlessly copying the buffer. (the `SharedBuffer` constructor does that)

lhofhansl replied:
Agreed. I followed what we had for zlib. Would have to change that as well. Lemme take a stab at it.

lhofhansl replied:
Won't I then have the reverse problem: Creating a string object, which makes a copy?
At best there could be a version that take a char[] and size, and both the string and sharedbuffer version call that one.

lhofhansl replied:
Did that... New push coming soon.

sfan5 reviewed (approved):
Looks good, just a few minor things.

sfan5 commented:
Would be nice to see the (de-)compression statistics code working with this.

Also since the opportunity exists now: 
* Should any fields be kept out of the compressed blob so they could be easily modified? Is there need for this?
* Are there any other improvements that could be made to the mapblock format? (e.g. from #3946)

lhofhansl commented:
There's also an option to change the "disk" format (the table schema). Often only the mapblock's time is updated, and yet the entire block has to be serialized and written again. If the timestamp would be a separate column in the schema one could just update that.

kilbith commented:
Zstd 1.5 released with big performance improvements: https://github.com/facebook/zstd/releases/tag/v1.5.0

lhofhansl commented:
I can't even use the 1.4.x APIs because we do not have that version everywhere. From I can gleam from the description the perf improvements are for inputs 256KB or larger. Here we are encrypting a few hundred bytes.

lhofhansl committed:
avoid copy during compression

lhofhansl commented:
Last part to do is that to add the benchmarking...

kilbith commented:
Why bothering with the benchmarking when we already know that zstd is inherently faster than zlib in all situations ?

lhofhansl commented:
I tried for quite a while to get the bench-marking code to work. Server has changed, etc. Saving outside of a map seems to not work, getting unknown id mapping, etc. I'll try more when I get time again.

sfan5 commented:
Do you have the commits anywhere so I can take a look?

lhofhansl commented:
@sfan5 Sorry forgot about it. Put them here: https://github.com/lhofhansl/minetest/tree/zstd-with-bench

lhofhansl commented:
Rebased...

lhofhansl commented:
Hmm... Android build still failing.

lhofhansl committed:
Android fix

lhofhansl commented:
OK... Android is fixed too. Thanks @sfan5 for doing the heavy lifting of the zstd port. Now this just needs the benchmarking to work.

lhofhansl commented:
Not sure what I tested before. Recompress/Decompresstimes is working fine. Please test! :)

lhofhansl committed:
Recompress/Decompresstimes

ExeVirus commented:
I see that this is getting quite close to a merge-able state, which is exciting! Have we considered the difference a dictionary could make in these compression sizes and speeds? Typically dictionaries really shine on small sized data like mapblocks, because normal headless compression merely looks at the given data block and compresses from that, whereas mapblocks in minetest are VERY often self-similar, and a dictionary should result in a 30-50% improvement in size and 20-30% in speed, based on my prior experience.

The other zstd dictionary use case would be with allowing servers to provide their own dictionary for their packet traffic after a client is loaded. typically can reduce server-client bandwidth required by 10% or so, but adds a slight amount of delay in processing, so it's more of a tradeoff there.

lhofhansl commented:
@ExeVirus The improvement in compression _and_ speed are already pretty good. I'd rather not muddle this with more changes.

You're not wrong about the dictionaries. We can pre-code one. Or we can leave the compressor open, and every now and then reset the dictionary.
As for transferring media... We'd have to offset that with extra CPU used on server and client. 

Also a question for everyone here: Should we remove the configurable compression-level now? Seems it's somewhat pointless with zstd, and one less config to worry about.

lhofhansl commented:
I'm not planning to do anything more here.

lhofhansl commented:
rebased after docker and android changes.

lhofhansl commented:
Rebased again after build changes.

Let's (1) merge, (2) suggest changes, or (3) close.

SmallJoker commented on src/mapblock.cpp:
```diff
@@ -381,38 +383,46 @@ void MapBlock::serialize(std::ostream &os, u8 version, bool disk, int compressio
+	if (version >= 29) {
+		m_node_metadata.serialize(os, version, disk);
+	} else {
+		// use oss from above to avoid allocating another stream object
```
That would be fine. Stream objects are initialized very quickly and this sort of optimization does not seem to be necessary here. It also makes the purpose of `oss` very obscure, because they're for two separate things.

SmallJoker replied:
Also this compression step is not task of `MapBlock` and should be moved to `NodeMetadata::serialize`.

lhofhansl replied:
I found creation of string streams to be very slow.

SmallJoker replied:
https://github.com/minetest/minetest/pull/10788#issuecomment-756967513

Hmm apparently. I wonder what's the cause for this :thinking:. Recycling these streams makes sense, but without proper naming and call structure (like suggested for MapNode in another comment) this results in code that's more difficult to understand and work with in the future.

lhofhansl replied:
Apparently it has to do locale handling upon creation.

sfan5 replied:
That sounds like it shouldn't apply to binary stringstreams, should it? :thinking:

lhofhansl replied:
There's no way I could find speeding up stringstream creation.

SmallJoker commented on src/mapblock.cpp:
```diff
@@ -381,38 +383,46 @@ void MapBlock::serialize(std::ostream &os, u8 version, bool disk, int compressio
 	/*
 		Bulk node data
 	*/
+	u8 content_width = 2;
+	u8 params_width = 2;
```
Should be `const`

lhofhansl replied:
Copied from existing code... But I agree.

lhofhansl replied:
done

sfan5 commented on src/main.cpp:
```diff
@@ -1023,3 +1036,134 @@ static bool migrate_map_database(const GameParams &game_params, const Settings &
+		MapBlock mb(NULL, *it, &server);
+
+		clock_t t = clock();
+		mb.deSerialize(iss, ver, true);
```
This also counts all time spent deserializing

SmallJoker commented on src/mapblock.cpp:
```diff
@@ -365,6 +365,8 @@ void MapBlock::serialize(std::ostream &os, u8 version, bool disk, int compressio
 	FATAL_ERROR_IF(version < SER_FMT_VER_LOWEST_WRITE, "Serialisation version error");
 
+	std::ostringstream oss(std::ios_base::binary);
+	std::ostream &os = version >= 29 ? oss : out;
```
Rename suggestion:
 * `oss` -> `os_raw`
 * `out` -> `os_compressed`

SmallJoker commented on src/mapblock.cpp:
```diff
@@ -460,10 +475,17 @@ void MapBlock::deSerialize(std::istream &is, u8 version, bool disk)
+		// decompress the whole thing
+		decompress(in, iss, version);
+	}
+	std::istream &is = version >= 29 ? iss : in;
```
Rename suggestion:
 * `iss` -> `in_raw`
 * `in` -> `in_compressed`

sfan5 commented on src/main.cpp:
```diff
@@ -1023,3 +1036,134 @@ static bool migrate_map_database(const GameParams &game_params, const Settings &
+		}
+
+		std::istringstream iss(data, std::ios_base::binary);
+		compdata += data.size();
```
...and this is flawed because not all bytes of a mapblock are compressed data

lhofhansl replied:
This change now compressed the block as a whole. It should be fairly accurate.

sfan5 replied:
Sure, but that's not the case for blocks in the old format.

SmallJoker commented on src/mapblock.cpp:
```diff
@@ -484,8 +506,16 @@ void MapBlock::deSerialize(std::istream &is, u8 version, bool disk)
+		MapNode::deSerializeBulk(is, version, data, nodecount,
+			content_width, params_width);
+	} else {
+		// use iss from above to avoid allocating another stream object
```
Same here. Recycling objects that are for another purpose only makes this code less readable than it already is.

SmallJoker commented on src/mapgen/mg_schematic.h:
```diff
@@ -70,6 +70,7 @@ class Server;
 #define MTSCHEM_FILE_SIGNATURE 0x4d54534d // 'MTSM'
 #define MTSCHEM_FILE_VER_HIGHEST_READ  4
 #define MTSCHEM_FILE_VER_HIGHEST_WRITE 4
+#define MTSCHEM_MAPNODE_SER_FMT_VER    28 // remember to up versions above when increasing this
```
Unclear instructions. What's to "up"? Why? When must this be increased? Why can't we use the map serialization version for schems as well to reduce dependencies?

lhofhansl replied:
@sfan5 A better idea? I took this from your older branch.

sfan5 replied:
Schematics must still be written (and read) using Zlib according to the specification.
If you were to introduce Zstd schematics (by bumping the ver to > 28), the file format of schematics needs to be bumped too.

perhaps ` // Fixed serialization version for schematics since these still need to use Zlib` would be better

nerzhul commented on src/serialization.cpp:
```diff
@@ -240,8 +347,24 @@ void compress(const SharedBuffer<u8> &data, std::ostream &os, u8 version)
 	os.write((char*)&current_byte, 1);
 }
 
+void compress(const SharedBuffer<u8> &data, std::ostream &os, u8 version, int level)
```
this function must clearly have the compression type in mind as we have 2 types of compression here

sfan5 replied:
It does that automatically depending on version (we have an old, third compression type already)

SmallJoker commented on src/mapnode.cpp:
```diff
@@ -779,15 +772,10 @@ void MapNode::deSerializeBulk(std::istream &is, int version,
-		throw SerializationError("deSerializeBulkNodes: "
-				"decompress resulted in invalid size");
-	const u8 *databuf = reinterpret_cast<const u8*>(s.c_str());
+	// read data
```
We have a `version` parameter. The decompression of the old version is NOT the MapBlock's task. Instead, it should be done here like previously to keep the responsibilities of each function sane.

lhofhansl replied:
Well... The mapblock is now compressed as a whole (to get better compression, better use of the dictionary, and better performance)

SmallJoker replied:
Yes, for versions 29+ that's true. However, for older ones it would be a cleaner call structure (IMO) when the MapNode decompression is done in MapNode and not MapBlock.

lhofhansl replied:
I actually think it's better the way it is now. Node know how to (de)serialize themselves. But all (de)compression logic is in MapBlock (the thing that actually gets stored in the DB).

Why would decompression not be the MapBlock's task? Before this responsibility was spread, now it's all together, which I think is cleaner.

sfan5 commented:
* there's still a review comment here about an incorrect description https://github.com/minetest/minetest/pull/10788/files#r626609425
* you don't need to fix the benchmark comments, but in general I don't think we should merge the benchmark options into master
* I just remembered that this can be updated too: https://github.com/minetest/minetest/blob/master/doc/world_format.txt#L279

SmallJoker commented on src/serialization.cpp:
```diff
@@ -197,27 +198,133 @@ void decompressZlib(std::istream &is, std::ostream &os, size_t limit)
+		}
+	} while (ret != 0);
+
+	// Unget all the data that ZSTD_decompressStream didn't take
```
Q: Why is this necessary? Does zstd read too much of `is`?

The `SerializationError` below also seems vague. What does it mean? Perhaps something like `"cannot restore original data after stream. data is corrupted now"`?

lhofhansl replied:
Same reasoning as the zlib code above. Decompression might have read more data out of the stream into its cache than it ends up needing, in that case it is important to restore these bytes back into the stream.

SmallJoker commented on src/serialization.cpp:
```diff
@@ -197,27 +198,133 @@ void decompressZlib(std::istream &is, std::ostream &os, size_t limit)
 	if(version >= 11)
 	{
-		compressZlib(*data ,data.getSize(), os);
+		compressZlib(data, size, os, level);
```
https://github.com/madler/zlib/blob/cacf7f1d4e3d44d871b605da3b647f07d718623f/deflate.c#L299

level -1 results in a compression error. 0 should be used as default.

lhofhansl replied:
Huh? I don't follow. For zlib -1 results in default compression.

SmallJoker replied:
Oh yes right. `Z_DEFAULT_COMPRESSION` is `-1`, and `level` is properly limited in `Server::SendBlockNoLock`.
Resolved.

SmallJoker commented on src/serialization.h:
```diff
@@ -67,9 +67,9 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #define SER_FMT_VER_INVALID 255
 // Highest supported serialization version
-#define SER_FMT_VER_HIGHEST_READ 28
+#define SER_FMT_VER_HIGHEST_READ 29
```
Needs documentation above.

lhofhansl replied:
"Above"? Is there doc file I need to update?

SmallJoker replied:
No, I meant [this huge comment](https://github.com/minetest/minetest/blob/63fc728a84a5ba97240233ad1c5d94f1ade2deb1/src/serialization.h#L27-L66)  which needs a new line to document version 29.

lhofhansl commented:
Addressed most of the comments. Will think about more about how to make the (de)compression of older version less convoluted in the code,

lhofhansl committed:
Address review comments

lhofhansl commented:
Ping...

sfan5 commented:
Everything mentioned in https://github.com/minetest/minetest/pull/10788#issuecomment-867037613 still applies as far as I can see.

lhofhansl commented:
I'll do another pass.

lhofhansl commented:
Did another pass.

OK... Here's what's open:

1. The compression levels. Right now it still maps allows zstd level from [1-10]. I found the other levels not useful, it also maps nicely to the zlib levels. Leave it that way, or change?
2. Leave the benchmark code in or not?

This is now open for more than 1/2 year - my fault, too, for not having much time on this, but let's iterate to get it in.

lhofhansl commented:
Rebased.

lhofhansl commented:
Will continue to rebase only if there is interest to actually finish it.

sfan5 commented:
> OK... Here's what's open:
> 1. The compression levels. Right now it still maps allows zstd level from [1-10]. I found the other levels not useful, it also maps nicely to the zlib levels. Leave it that way, or change?

Should be left as it is.

> 2. Leave the benchmark code in or not?

My opinion (still) is to remove it, it's purely testing code and also because it doesn't give fully accurate results.
Originally I wanted to run the benchmark code myself to compare performance and see if the default compression level could/should be changed but looks like I won't get to that. That can be done at any later point after this PR is merged in fact.

sfan5 commented on doc/world_format.txt:
```diff
@@ -281,6 +282,8 @@ MapBlock serialization format
 NOTE: Byte order is MSB first (big-endian).
 NOTE: Zlib data is in such a format that Python's zlib at least can
       directly decompress.
+NOTE: In version 29 zstd is used instead of zlib. In addition the entire
+      block is first serialized and then compressed.
```
```suggestion
NOTE: Since version 29 zstd is used instead of zlib. In addition the entire
      block is first serialized and then compressed (except the version byte).
```

sfan5 commented on doc/world_format.txt:
```diff
@@ -335,7 +338,7 @@ u8 params_width
 - Always 2
 
-zlib-compressed node data:
+zlib-compressed node data (version <= 28):
```
```suggestion
node data (zlib-compressed if version <= 28):
```

sfan5 commented on doc/world_format.txt:
```diff
@@ -348,7 +351,7 @@ if content_width == 2:
 - The location of a node in each of those arrays is (z*16*16 + y*16 + x).
 
-zlib-compressed node metadata list
+zlib-compressed node metadata list (version <= 28)
```
```suggestion
node metadata list (zlib-compressed if version <= 28)
```

sfan5 commented:
Okay I've given this a test and it works fine. Also verified the map format by updating a script I have which fully parses a mapblock.
At this point the removal of the benchmarking code and the thing I'm about to mention are my only requests, :+1: otherwise.

...what I noticed however is this: Since map metadata is no longer in its' own zlib blob, the implicit length is gone which means you can no longer skip metadata without parsing it!
While this is of no concern of the engine, it is painful for external tools which could until now skip data they weren't interested in (fairly easily](https://github.com/minetest/minetestmapper/blob/fd4c5dd23253c8c8662cbec6fa898f76fe7d2e5a/BlockDecoder.cpp#L80-L104). Parsing the map metadata requires working through the variables and handling ASCII text to find the line that contains `EndInventory`.

How would I fix this? By writing the metadata using `writeLongString`, which adds a length field. This will easily allow skipping written metadata as before.

sfan5 commented on src/mapblock.cpp:
```diff
@@ -494,15 +524,20 @@ void MapBlock::deSerialize(std::istream &is, u8 version, bool disk)
+		}
 	} catch(SerializationError &e) {
 		warningstream<<"MapBlock::deSerialize(): Ignoring an error"
 				<<" while deserializing node metadata at ("
```
Btw this has to be made fatal instead of a warning on `ver >= 29` too because of what I mentioned.

lhofhansl replied:
How about the latest? Can't say I like it, though. It creates yet another ostringstream each time we safe a mapblock and there are *two* additional copy of the bytes representing the metadata.

sfan5 replied:
Fine by me.

lhofhansl committed:
Write length of metadata

lhofhansl committed:
Fix world format doc

lhofhansl committed:
remove benchmarking code

lhofhansl commented:
OK... Added the length marker for the metadata, fixed the world_format doc, removed the decompress-benchmark (but left the db recompression in), and rebased.

sfan5 commented on doc/world_format.txt:
```diff
@@ -348,7 +351,7 @@ if content_width == 2:
 - The location of a node in each of those arrays is (z*16*16 + y*16 + x).
 
-zlib-compressed node metadata list
+node metadata list (zlib-compressed if version <= 28)
```
```suggestion
if map format version >= 29:
  u32 length of all following bytes that belong to metadata
node metadata list (zlib-compressed if version <= 28)
```

sfan5 commented on src/mapblock.cpp:
```diff
@@ -494,15 +528,22 @@ void MapBlock::deSerialize(std::istream &is, u8 version, bool disk)
+					&m_node_metadata, &m_node_timers,
+					m_gamedef->idef());
+		}
 	} catch(SerializationError &e) {
```
```suggestion
	} catch(SerializationError &e) {
		if (version >= 29)
			throw e;
```

sfan5 commented on src/main.cpp:
```diff
@@ -1034,3 +1041,75 @@ static bool migrate_map_database(const GameParams &game_params, const Settings &
+
+	actionstream << "############" << std::endl;
+	actionstream << "Recompressed blocks:            " << count << std::endl;
+	actionstream << "Total data (before comp.):      " << (decompdata / 1024 / 1024) << " MB" << std::endl;
+	actionstream << "Total compression CPU time:     " << comptime << "s" << std::endl;
+	actionstream << "Compression CPU time per block: " << (comptime / count * 1000 * 1000) << "us" << std::endl;
+	actionstream << "Compression speed:              " << (decompdata / comptime / 1024) << " KB/s" << std::endl;
```
Also these stats should be removed, none of them are accurate.

lhofhansl commented:
Another idea for the meta-data length: Instead of writing a separate ostringstream, one could just leave a 4 byte placeholder, then remember the position before and after writing the metadata (`tellp()`) and then poke that value in the to std::string before it is written into the compressed stream.

sfan5 commented:
If you can make that work that'd be good but I thought ostringstreams did not support seeking.

lhofhansl commented:
According to the c++ standard `stringbuf` does support seeking.
And `tellp()` seems to work. Verified by comparing the size with the difference of the offsets.
(By the way I noticed that most cases the metadata size is just 1 - i.e. no metadata - in that case the extra ostringstream is incredibly wasteful)

I'll post an update soon.

lhofhansl commented:
Yet another alternative is to write the metadata last. That way it can simply be skipped by parsing out the rest (all fix sized) and then skipping to the end.

lhofhansl committed:
Determine metadata length without an extra ostringstream

lhofhansl commented:
Now the MacOS build is broken. :(

sfan5 commented:
> Yet another alternative is to write the metadata last.

Ahhhh of course why didn't I think of that, we can change the format freely anyway.
Sorry to make you do more but at this point I suggest the following format changes:
* Move the timestamp after `lighting_complete` (not very relevant but who knows when it'll be useful)
* Move the node/id mapping just before `content_width`
* Remove the metadata length field again

This is even easier for minetestmapper and similar tools because they can just stop reading before metadata, static objects or anything else that is uninteresting.

lhofhansl commented:
Thanks @sfan5 . Update coming soon :)

lhofhansl committed:
Rearrange serialization order

lhofhansl committed:
(hopefully) fix MacOS build

lhofhansl committed:
Remove last benchmark code

lhofhansl commented:
I still need to update world_format.txt.

lhofhansl committed:
update world_format.txt

lhofhansl commented:
Alternatively we would document version >= 29 separately and leave <= 28 the way it is.

sfan5 commented on src/main.cpp:
```diff
@@ -1034,3 +1041,65 @@ static bool migrate_map_database(const GameParams &game_params, const Settings &
+	std::cerr << std::endl;
+	db->endSave();
+
+	actionstream << "############" << std::endl;
+	actionstream << "Recompressed blocks:            " << count << std::endl;
+	actionstream << "############" << std::endl;
```
```suggestion
	actionstream << "Done, " << count << " blocks were recompressed." << std::endl;
```

sfan5 commented on src/mapblock.cpp:
```diff
@@ -549,6 +613,18 @@ void MapBlock::deSerialize(std::istream &is, u8 version, bool disk)
+				<<" while deserializing node metadata at ("
+				<<PP(getPos())<<": "<<e.what()<<std::endl;
+		}
+	}
```
Hm, so metadata is now at the end?
Not that it really matters but I expected it to remain where it was before.

lhofhansl replied:
Oh... I thought that's what we discussed. It's at the end so that you can ignore it. Maybe I misunderstood. I don't care whether it's at the end, as long as it works. :)

sfan5 replied:
Well with the node-id mapping now moved before the node data you can simply stop reading there, no need for the meta to be moved.
If you don't mind I could checkout your branch and push a commit moving it back.

lhofhansl replied:
I'll do it. NP.

sfan5 commented on src/serialization.h:
```diff
@@ -63,13 +63,14 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 	26: Never written; read the same as 25
 	27: Added light spreading flags to blocks
 	28: Added "private" flag to NodeMetadata
+	29: Switched compression to zstd
```
```suggestion
	29: Switched compression to zstd, a bit of reorganization
```

sfan5 committed:
move meta back

sfan5 committed:
apply other suggestions

sfan5 reviewed (approved):
Works, tested:
* old map loadability
* network compatibility, both ways
* format conforming to expectations

sfan5 committed:
minor things

lhofhansl commented:
Changes look good.
